src/libcore/iter.rs

```diff
@@ -1024,6 +1014,38 @@ pub trait Iterator {
     }
 }
 
+/// Select an element from an iterator based on the given projection
+/// and "comparison" function.
+///
+/// This is an idiosyncratic helper to try to factor out the
+/// commonalities of {max,min}{,_by}. In particular, this avoids
+/// having to implement optimisations several times.
+#[inline]
+fn select_fold1<I,B, FProj, FCmp>(mut it: I,
+                                  mut f_proj: FProj,
+                                  mut f_cmp: FCmp) -> Option<(B, I::Item)>
+    where I: Iterator,
+          FProj: FnMut(&I::Item) -> B,
+          FCmp: FnMut(&B, &I::Item, &B, &I::Item) -> bool
```

Comment:
Is there some reason the comparison function takes 4 arguments, when each client only wants 2 of them, instead of just making it take the projections and having `max()` / `min()` use `|x| x` for the projection function?